### PR TITLE
Allow consecutive signatures in different terms

### PR DESCRIPTION
**Pull request by achamayou**
Spotted in #5973, the current constraint seems too strong. It is expected that we will see consecutive signatures after elections, just in different terms.

Also disable CheckQuorum in the reconfiguration MC configs to keep model checking time under control.

### CODE REVIEW

**Comment by achamayou** [comment]
Two things are interesting here:

1. Turning off CheckQuorum brings model checking time back in line to what existed before the signature changes
2. It causes quite a few trace validation failures, not all of them obviously correct (reelection in particular)

I will look at 2. tomorrow, but insights welcome on 1.

**Comment by lemmy** [comment]
Number if *distinct* variables values without and with https://github.com/microsoft/CCF/pull/6041/commits/ea4ceba555791c82094b111675e8f0cf678acbff after 1.1 million distinct states

```
Progress(40) at 2024-03-04 10:29:02: 2,844,448 states generated (484,724 s/min), 1,191,966 distinct states found (219,208 ds/min), 100,165 states left on queue.
removedFromConfiguration: 2
leadershipState: 11
matchIndex: 25
votedFor: 6
commitIndex: 34
membershipState: 4
messages: 13037
isNewFollower: 1
currentTerm: 20
log: 5163
sentIndex: 98
votesGranted: 9
configurations: 33
```

```
Progress(25) at 2024-03-04 10:46:38: 2,502,858 states generated (648,259 s/min), 1,139,891 distinct states found (267,189 ds/min), 286,174 states left on queue.
removedFromConfiguration: 2
leadershipState: 9
matchIndex: 6
votedFor: 6
commitIndex: 30
membershipState: 4
messages: 7549
isNewFollower: 1
currentTerm: 18
log: 8803
sentIndex: 108
votesGranted: 6
configurations: 24
```

`messages`, `log`, and `matchIndex` most notable differences.

**Comment by achamayou** [comment]
@lemmy thank you for those, what stands out to me is that with https://github.com/microsoft/CCF/commit/ea4ceba555791c82094b111675e8f0cf678acbff, we seem to have many more distinct log values, roughly twice as many, with only half as many distinct message values.

This is not unexpected, because by forbidding the emission of a signature at the beginning of a term, we depend on a client request being emitted first, followed by a signature, to reach commit. That's more opportunities for timeouts (and checkquorums), and so more messages , with less log progress, seems intuitive.

It doesn't give a very good intuition of why the model checking takes dramatically longer though.

**Comment by lemmy** [comment]
https://github.com/microsoft/CCF/commit/ea4ceba555791c82094b111675e8f0cf678acbff finishes in a little more than an hour on a 32 core Codespace:

```
/home/codespace/java/current//bin/java -XX:+UseParallelGC -Dtlc2.tool.impl.Tool.cdot=true -cp /home/codespace/.vscode-remote/extensions/alygin.vscode-tlaplus-nightly-2024.3.306/tools/tla2tools.jar -DTLA-Library= tlc2.TLC MCccfraft.tla -tool -modelcheck -config MCccfraftAtomicReconfig.cfg -workers auto

TLC2 Version 2.18 of Day Month 20?? (rev: 2576f2c)
Running breadth-first search Model-Checking with fp 0 and seed 2126674445793187749 with 32 workers on 32 cores with 27305MB heap and 64MB offheap memory [pid: 6083] (Linux 6.2.0-1019-azure amd64, Microsoft 21.0.2 x86_64, MSBDiskFPSet, DiskStateQueue).
Starting SANY...
Parsing file /workspaces/CCF/tla/consensus/MCccfraft.tla
Parsing file /workspaces/CCF/tla/consensus/ccfraft.tla
Parsing file /workspaces/CCF/tla/consensus/StatsFile.tla
Parsing file /workspaces/CCF/tla/consensus/MCAliases.tla
Parsing file /tmp/tlc-17579067739641887089/Naturals.tla (jar:file:/home/codespace/.vscode-remote/extensions/alygin.vscode-tlaplus-nightly-2024.3.306/tools/tla2tools.jar!/tla2sany/StandardModules/Naturals.tla)
Parsing file /tmp/tlc-17579067739641887089/FiniteSets.tla (jar:file:/home/codespace/.vscode-remote/extensions/alygin.vscode-tlaplus-nightly-2024.3.306/tools/tla2tools.jar!/tla2sany/StandardModules/FiniteSets.tla)
Parsing file /tmp/tlc-17579067739641887089/Sequences.tla (jar:file:/home/codespace/.vscode-remote/extensions/alygin.vscode-tlaplus-nightly-2024.3.306/tools/tla2tools.jar!/tla2sany/StandardModules/Sequences.tla)
Parsing file /tmp/tlc-17579067739641887089/TLC.tla (jar:file:/home/codespace/.vscode-remote/extensions/alygin.vscode-tlaplus-nightly-2024.3.306/tools/tla2tools.jar!/tla2sany/StandardModules/TLC.tla)
Parsing file /tmp/tlc-17579067739641887089/FiniteSetsExt.tla (jar:file:/home/codespace/.vscode-remote/extensions/alygin.vscode-tlaplus-nightly-2024.3.306/tools/CommunityModules-deps.jar!/FiniteSetsExt.tla)
Parsing file /tmp/tlc-17579067739641887089/SequencesExt.tla (jar:file:/home/codespace/.vscode-remote/extensions/alygin.vscode-tlaplus-nightly-2024.3.306/tools/CommunityModules-deps.jar!/SequencesExt.tla)
Parsing file /tmp/tlc-17579067739641887089/Functions.tla (jar:file:/home/codespace/.vscode-remote/extensions/alygin.vscode-tlaplus-nightly-2024.3.306/tools/CommunityModules-deps.jar!/Functions.tla)
Parsing file /tmp/tlc-17579067739641887089/Folds.tla (jar:file:/home/codespace/.vscode-remote/extensions/alygin.vscode-tlaplus-nightly-2024.3.306/tools/CommunityModules-deps.jar!/Folds.tla)
Parsing file /tmp/tlc-17579067739641887089/Json.tla (jar:file:/home/codespace/.vscode-remote/extensions/alygin.vscode-tlaplus-nightly-2024.3.306/tools/tla2tools.jar!/tla2sany/StandardModules/Json.tla)
Parsing file /workspaces/CCF/tla/consensus/Network.tla
Parsing file /tmp/tlc-17579067739641887089/Bags.tla (jar:file:/home/codespace/.vscode-remote/extensions/alygin.vscode-tlaplus-nightly-2024.3.306/tools/tla2tools.jar!/tla2sany/StandardModules/Bags.tla)
Parsing file /tmp/tlc-17579067739641887089/TLCExt.tla (jar:file:/home/codespace/.vscode-remote/extensions/alygin.vscode-tlaplus-nightly-2024.3.306/tools/tla2tools.jar!/tla2sany/StandardModules/TLCExt.tla)
Parsing file /tmp/tlc-17579067739641887089/Integers.tla (jar:file:/home/codespace/.vscode-remote/extensions/alygin.vscode-tlaplus-nightly-2024.3.306/tools/tla2tools.jar!/tla2sany/StandardModules/Integers.tla)
Semantic processing of module Naturals
Semantic processing of module Sequences
Semantic processing of module FiniteSets
Semantic processing of module TLC
Semantic processing of module Folds
Semantic processing of module Functions
Semantic processing of module FiniteSetsExt
Semantic processing of module SequencesExt
Semantic processing of module Bags
Semantic processing of module Network
Semantic processing of module ccfraft
Semantic processing of module Json
Semantic processing of module StatsFile
Semantic processing of module Integers
Semantic processing of module TLCExt
Semantic processing of module MCAliases
Semantic processing of module MCccfraft
SANY finished.
Starting... (2024-03-04 21:37:38)
Computing initial states...
Finished computing initial states: 1 distinct state generated at 2024-03-04 21:37:38.
Progress(20) at 2024-03-04 21:37:41: 139,379 states generated (139,379 s/min), 77,105 distinct states found (77,105 ds/min), 26,637 states left on queue.
Progress(33) at 2024-03-04 21:38:41: 8,376,920 states generated (8,237,541 s/min), 3,200,799 distinct states found (3,123,694 ds/min), 409,658 states left on queue.
Progress(40) at 2024-03-04 21:39:41: 15,752,866 states generated (7,375,946 s/min), 6,068,132 distinct states found (2,867,333 ds/min), 490,739 states left on queue.
Progress(44) at 2024-03-04 21:40:41: 22,237,583 states generated (6,484,717 s/min), 8,845,156 distinct states found (2,777,024 ds/min), 703,919 states left on queue.
Progress(47) at 2024-03-04 21:41:41: 28,390,336 states generated (6,152,753 s/min), 11,485,892 distinct states found (2,640,736 ds/min), 1,148,171 states left on queue.
Progress(49) at 2024-03-04 21:42:41: 34,563,055 states generated (6,172,719 s/min), 14,066,171 distinct states found (2,580,279 ds/min), 1,629,824 states left on queue.
Progress(50) at 2024-03-04 21:43:41: 40,845,288 states generated (6,282,233 s/min), 16,494,444 distinct states found (2,428,273 ds/min), 2,126,668 states left on queue.
Progress(51) at 2024-03-04 21:44:41: 47,142,303 states generated (6,297,015 s/min), 18,904,981 distinct states found (2,410,537 ds/min), 2,568,282 states left on queue.
Progress(52) at 2024-03-04 21:45:41: 53,335,858 states generated (6,193,555 s/min), 21,277,235 distinct states found (2,372,254 ds/min), 3,000,045 states left on queue.
Progress(53) at 2024-03-04 21:46:41: 59,751,929 states generated (6,416,071 s/min), 23,560,338 distinct states found (2,283,103 ds/min), 3,376,841 states left on queue.
Progress(53) at 2024-03-04 21:47:41: 66,260,873 states generated (6,508,944 s/min), 25,787,568 distinct states found (2,227,230 ds/min), 3,777,085 states left on queue.
Progress(54) at 2024-03-04 21:48:41: 72,495,428 states generated (6,234,555 s/min), 28,068,823 distinct states found (2,281,255 ds/min), 4,140,493 states left on queue.
Progress(54) at 2024-03-04 21:49:41: 79,134,589 states generated (6,639,161 s/min), 30,222,501 distinct states found (2,153,678 ds/min), 4,479,017 states left on queue.
Progress(55) at 2024-03-04 21:50:41: 85,512,671 states generated (6,378,082 s/min), 32,449,332 distinct states found (2,226,831 ds/min), 4,782,542 states left on queue.
Progress(55) at 2024-03-04 21:51:41: 91,968,073 states generated (6,455,402 s/min), 34,608,804 distinct states found (2,159,472 ds/min), 5,130,565 states left on queue.
Progress(55) at 2024-03-04 21:52:41: 98,777,211 states generated (6,809,138 s/min), 36,675,048 distinct states found (2,066,244 ds/min), 5,358,895 states left on queue.
Progress(56) at 2024-03-04 21:53:41: 104,928,810 states generated (6,151,599 s/min), 38,914,291 distinct states found (2,239,243 ds/min), 5,688,223 states left on queue.
Progress(56) at 2024-03-04 21:54:41: 111,717,610 states generated (6,788,800 s/min), 40,970,268 distinct states found (2,055,977 ds/min), 5,893,301 states left on queue.
Progress(56) at 2024-03-04 21:55:41: 118,447,401 states generated (6,729,791 s/min), 42,989,295 distinct states found (2,019,027 ds/min), 6,077,457 states left on queue.
Progress(57) at 2024-03-04 21:56:41: 124,551,579 states generated (6,104,178 s/min), 45,194,942 distinct states found (2,205,647 ds/min), 6,398,295 states left on queue.
Progress(57) at 2024-03-04 21:57:41: 131,355,305 states generated (6,803,726 s/min), 47,208,724 distinct states found (2,013,782 ds/min), 6,563,847 states left on queue.
Progress(57) at 2024-03-04 21:58:41: 138,213,460 states generated (6,858,155 s/min), 49,176,286 distinct states found (1,967,562 ds/min), 6,696,932 states left on queue.
Progress(58) at 2024-03-04 21:59:41: 144,474,358 states generated (6,260,898 s/min), 51,318,521 distinct states found (2,142,235 ds/min), 6,916,077 states left on queue.
Progress(58) at 2024-03-04 22:00:42: 150,927,910 states generated (6,453,552 s/min), 53,393,966 distinct states found (2,075,445 ds/min), 7,146,896 states left on queue.
Progress(58) at 2024-03-04 22:01:42: 157,969,445 states generated (7,041,535 s/min), 55,333,365 distinct states found (1,939,399 ds/min), 7,243,432 states left on queue.
Progress(58) at 2024-03-04 22:02:42: 164,645,938 states generated (6,676,493 s/min), 57,332,348 distinct states found (1,998,983 ds/min), 7,332,696 states left on queue.
Progress(59) at 2024-03-04 22:03:42: 170,867,322 states generated (6,221,384 s/min), 59,420,692 distinct states found (2,088,344 ds/min), 7,548,250 states left on queue.
Progress(59) at 2024-03-04 22:04:42: 177,521,138 states generated (6,653,816 s/min), 61,417,478 distinct states found (1,996,786 ds/min), 7,707,216 states left on queue.
Progress(59) at 2024-03-04 22:05:42: 184,691,344 states generated (7,170,206 s/min), 63,304,835 distinct states found (1,887,357 ds/min), 7,703,837 states left on queue.
Progress(59) at 2024-03-04 22:06:42: 191,170,232 states generated (6,478,888 s/min), 65,321,310 distinct states found (2,016,475 ds/min), 7,803,121 states left on queue.
Checkpointing of run states/24-03-04-21-37-38.041
Checkpointing completed at (2024-03-04 22:07:42)
Progress(60) at 2024-03-04 22:07:42: 197,518,242 states generated (6,348,010 s/min), 67,340,727 distinct states found (2,019,417 ds/min), 7,963,390 states left on queue.
Progress(60) at 2024-03-04 22:08:42: 204,174,736 states generated (6,656,494 s/min), 69,301,628 distinct states found (1,960,901 ds/min), 8,075,991 states left on queue.
Progress(60) at 2024-03-04 22:09:42: 211,335,473 states generated (7,160,737 s/min), 71,163,053 distinct states found (1,861,425 ds/min), 8,042,659 states left on queue.
Progress(60) at 2024-03-04 22:10:42: 218,029,830 states generated (6,694,357 s/min), 73,099,323 distinct states found (1,936,270 ds/min), 8,061,434 states left on queue.
Progress(61) at 2024-03-04 22:11:42: 224,396,081 states generated (6,366,251 s/min), 75,111,266 distinct states found (2,011,943 ds/min), 8,173,975 states left on queue.
Progress(61) at 2024-03-04 22:12:42: 230,906,874 states generated (6,510,793 s/min), 77,040,630 distinct states found (1,929,364 ds/min), 8,276,786 states left on queue.
Progress(61) at 2024-03-04 22:13:42: 238,006,394 states generated (7,099,520 s/min), 78,902,330 distinct states found (1,861,700 ds/min), 8,225,281 states left on queue.
Progress(61) at 2024-03-04 22:14:42: 244,954,915 states generated (6,948,521 s/min), 80,755,429 distinct states found (1,853,099 ds/min), 8,162,832 states left on queue.
Progress(62) at 2024-03-04 22:15:42: 251,167,587 states generated (6,212,672 s/min), 82,764,687 distinct states found (2,009,258 ds/min), 8,260,547 states left on queue.
Progress(62) at 2024-03-04 22:16:42: 257,776,578 states generated (6,608,991 s/min), 84,650,073 distinct states found (1,885,386 ds/min), 8,319,966 states left on queue.
Progress(62) at 2024-03-04 22:17:42: 264,718,744 states generated (6,942,166 s/min), 86,510,718 distinct states found (1,860,645 ds/min), 8,264,643 states left on queue.
Progress(62) at 2024-03-04 22:18:42: 271,831,064 states generated (7,112,320 s/min), 88,322,522 distinct states found (1,811,804 ds/min), 8,119,994 states left on queue.
Progress(62) at 2024-03-04 22:19:42: 278,110,087 states generated (6,279,023 s/min), 90,287,732 distinct states found (1,965,210 ds/min), 8,143,123 states left on queue.
Progress(63) at 2024-03-04 22:20:42: 284,774,557 states generated (6,664,470 s/min), 92,161,595 distinct states found (1,873,863 ds/min), 8,159,300 states left on queue.
Progress(63) at 2024-03-04 22:21:42: 291,503,785 states generated (6,729,228 s/min), 93,989,965 distinct states found (1,828,370 ds/min), 8,102,455 states left on queue.
Progress(63) at 2024-03-04 22:22:42: 298,800,972 states generated (7,297,187 s/min), 95,751,694 distinct states found (1,761,729 ds/min), 7,895,927 states left on queue.
Progress(63) at 2024-03-04 22:23:42: 305,212,409 states generated (6,411,437 s/min), 97,650,548 distinct states found (1,898,854 ds/min), 7,872,560 states left on queue.
Progress(64) at 2024-03-04 22:24:42: 311,828,292 states generated (6,615,883 s/min), 99,516,487 distinct states found (1,865,939 ds/min), 7,811,148 states left on queue.
Progress(64) at 2024-03-04 22:25:42: 318,492,471 states generated (6,664,179 s/min), 101,343,613 distinct states found (1,827,126 ds/min), 7,759,326 states left on queue.
Progress(64) at 2024-03-04 22:26:42: 325,769,420 states generated (7,276,949 s/min), 103,061,739 distinct states found (1,718,126 ds/min), 7,505,882 states left on queue.
Progress(64) at 2024-03-04 22:27:42: 332,308,885 states generated (6,539,465 s/min), 104,905,845 distinct states found (1,844,106 ds/min), 7,427,444 states left on queue.
Progress(65) at 2024-03-04 22:28:42: 338,889,685 states generated (6,580,800 s/min), 106,751,972 distinct states found (1,846,127 ds/min), 7,316,280 states left on queue.
Progress(65) at 2024-03-04 22:29:42: 345,579,688 states generated (6,690,003 s/min), 108,547,874 distinct states found (1,795,902 ds/min), 7,217,626 states left on queue.
Progress(65) at 2024-03-04 22:30:42: 352,899,155 states generated (7,319,467 s/min), 110,254,203 distinct states found (1,706,329 ds/min), 6,925,398 states left on queue.
Progress(65) at 2024-03-04 22:31:42: 359,195,839 states generated (6,296,684 s/min), 112,093,459 distinct states found (1,839,256 ds/min), 6,813,752 states left on queue.
Progress(66) at 2024-03-04 22:32:42: 365,920,858 states generated (6,725,019 s/min), 113,859,014 distinct states found (1,765,555 ds/min), 6,663,883 states left on queue.
Progress(66) at 2024-03-04 22:33:42: 372,828,287 states generated (6,907,429 s/min), 115,583,730 distinct states found (1,724,716 ds/min), 6,438,894 states left on queue.
Progress(66) at 2024-03-04 22:34:42: 379,839,097 states generated (7,010,810 s/min), 117,295,178 distinct states found (1,711,448 ds/min), 6,151,671 states left on queue.
Progress(67) at 2024-03-04 22:35:42: 386,307,293 states generated (6,468,196 s/min), 119,073,347 distinct states found (1,778,169 ds/min), 5,978,683 states left on queue.
Progress(67) at 2024-03-04 22:36:42: 393,041,718 states generated (6,734,425 s/min), 120,804,663 distinct states found (1,731,316 ds/min), 5,776,616 states left on queue.
Checkpointing of run states/24-03-04-21-37-38.041
Checkpointing completed at (2024-03-04 22:37:46)
Progress(67) at 2024-03-04 22:37:46: 400,574,646 states generated (7,532,928 s/min), 122,579,373 distinct states found (1,774,710 ds/min), 5,401,763 states left on queue.
Progress(68) at 2024-03-04 22:38:46: 407,073,658 states generated (6,499,012 s/min), 124,341,530 distinct states found (1,762,157 ds/min), 5,181,399 states left on queue.
Progress(68) at 2024-03-04 22:39:46: 413,905,530 states generated (6,831,872 s/min), 126,026,581 distinct states found (1,685,051 ds/min), 4,900,857 states left on queue.
Progress(68) at 2024-03-04 22:40:46: 420,763,767 states generated (6,858,237 s/min), 127,688,483 distinct states found (1,661,902 ds/min), 4,559,362 states left on queue.
Progress(69) at 2024-03-04 22:41:46: 427,448,176 states generated (6,684,409 s/min), 129,386,342 distinct states found (1,697,859 ds/min), 4,250,714 states left on queue.
Progress(69) at 2024-03-04 22:42:46: 434,404,037 states generated (6,955,861 s/min), 131,029,026 distinct states found (1,642,684 ds/min), 3,856,927 states left on queue.
Progress(70) at 2024-03-04 22:43:46: 440,980,664 states generated (6,576,627 s/min), 132,715,189 distinct states found (1,686,163 ds/min), 3,522,656 states left on queue.
Progress(70) at 2024-03-04 22:44:46: 447,825,290 states generated (6,844,626 s/min), 134,320,651 distinct states found (1,605,462 ds/min), 3,105,395 states left on queue.
Progress(71) at 2024-03-04 22:45:46: 454,477,719 states generated (6,652,429 s/min), 135,959,499 distinct states found (1,638,848 ds/min), 2,701,595 states left on queue.
Progress(72) at 2024-03-04 22:46:46: 461,176,877 states generated (6,699,158 s/min), 137,576,627 distinct states found (1,617,128 ds/min), 2,229,539 states left on queue.
Progress(73) at 2024-03-04 22:47:46: 467,881,972 states generated (6,705,095 s/min), 139,167,621 distinct states found (1,590,994 ds/min), 1,731,493 states left on queue.
Progress(74) at 2024-03-04 22:48:46: 474,472,949 states generated (6,590,977 s/min), 140,723,977 distinct states found (1,556,356 ds/min), 1,188,617 states left on queue.
Progress(76) at 2024-03-04 22:49:46: 480,998,098 states generated (6,525,149 s/min), 142,269,071 distinct states found (1,545,094 ds/min), 557,411 states left on queue.
"Writing stats to file: MCccfraftAtomicReconfig_stats.json"
Model checking completed. No error has been found.
  Estimates of the probability that TLC did not check all reachable states
  because two distinct states had the same fingerprint:
  calculated (optimistic):  val = .0027
  based on the actual fingerprints:  val = .0011
Progress(92) at 2024-03-04 22:50:38: 485,385,407 states generated (6,648,816 s/min), 143,443,802 distinct states found (1,964,895 ds/min), 0 states left on queue.
485385407 states generated, 143443802 distinct states found, 0 states left on queue.
The depth of the complete state graph search is 92.
The average outdegree of the complete state graph is 1 (minimum is 0, the maximum 7 and the 95th percentile is 3).
Finished in 4381624ms at (2024-03-04 22:50:39)
```

**Review by eddyashton — APPROVED** [review]

File: tla/consensus/MCccfraftAtomicReconfig.cfg [hunk 1]
```diff
@@ -14,6 +14,7 @@ CONSTANTS
     ClientRequest <- MCClientRequest
     SignCommittableMessages <- MCSignCommittableMessages
     ChangeConfigurationInt <- MCChangeConfigurationInt
+    CheckQuorum <- MCCheckQuorum
```
[the remark this answers is not in the document]
Comment (achamayou) [reply tla/consensus/MCccfraftAtomicReconfig.cfg:17]: @achamayou it does not seem to work:

```
Error: Successor state is not completely specified by action CheckQuorum of the next-state relation. The following variables are not assigned: commitIndex, configurations, currentTerm, isNewFollower, leadershipState, log, matchIndex, membershipState, messages, removedFromConfiguration, sentIndex, votedFor, votesGranted.
```

**Pull request merged**